File: WoofIOS/SwiftPackages/Presentation/Favorites/Sources/PresentationFavorites/UI/FavoritesScreen.swift
import DesignSystem
import PresentationCommon
import Resources
import SwiftUI

public struct FavoritesScreen: View {
    private let state: FavoritesViewModel.State
    private let onCellTap: (FavoriteCellUIModel) -> Void

    public init(state: FavoritesViewModel.State, onCellTap: @escaping (FavoriteCellUIModel) -> Void) {
        self.state = state
        self.onCellTap = onCellTap
    }

    public var body: some View {
        switch state {
        case .empty:
            FavoritesEmptyScreen()
        case .loaded(let cells):
            FavoritesLoadedScreen(cells: cells, onCellTap: onCellTap)
        }
    }
}

private struct FavoritesEmptyScreen: View {
    var body: some View {
        TemplateCenteredContent(topBar: { OrgNavigationHeaderTitle(title: L10n.Favorites.title) }) {
            OrgEmptyState(
                title: L10n.Favorites.emptyTitle,
                message: L10n.Favorites.emptyMessage
            )
        }
    }
}

private struct FavoritesLoadedScreen: View {
    let cells: [FavoriteCellUIModel]
    let onCellTap: (FavoriteCellUIModel) -> Void

    var body: some View {
        TemplateGrid(topBar: { OrgNavigationHeaderTitle(title: L10n.Favorites.title) }) {
            ForEach(cells) { cell in
                AsyncImageStateProvider(url: cell.photoUrl) { imageState in
                    OrgPhotoCard(
                        title: cell.name,
                        imageState: imageState,
                        contentDescription: L10n.Accessibility.dogPhoto(cell.name),
                        onTap: { onCellTap(cell) }
                    )
                }
            }
        }
    }
}

#Preview("Loaded") {
    ThemedScreenPreview(theme: WoofTheme.light()) {
        FavoritesScreen(state: FavoritesPreviewData.loaded(), onCellTap: { _ in })
    }
}

#Preview("Empty") {
    ThemedScreenPreview(theme: WoofTheme.dark()) {
        FavoritesScreen(state: .empty, onCellTap: { _ in })
    }
}

File: WoofAndroid/presentation/favorites/src/main/kotlin/com/perrystreet/woof/presentation/favorites/ui/FavoritesScreen.kt
package com.perrystreet.woof.presentation.favorites.ui

import androidx.compose.foundation.lazy.grid.items
import androidx.compose.runtime.Composable
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.PreviewParameter
import com.perrystreet.woof.designsystem.atomic.organisms.card.OrgPhotoCard
import com.perrystreet.woof.designsystem.atomic.organisms.emptystate.OrgEmptyState
import com.perrystreet.woof.designsystem.atomic.organisms.header.OrgNavigationHeaderTitle
import com.perrystreet.woof.designsystem.atomic.templates.TemplateCenteredContent
import com.perrystreet.woof.designsystem.atomic.templates.TemplateGrid
import com.perrystreet.woof.designsystem.preview.PreviewDevices
import com.perrystreet.woof.designsystem.preview.ThemeProvider
import com.perrystreet.woof.designsystem.preview.ThemedScreenPreview
import com.perrystreet.woof.designsystem.theme.ITheme
import com.perrystreet.woof.presentation.common.image.AsyncImageStateExtensions.rememberAsyncImageState
import com.perrystreet.woof.presentation.favorites.ui.preview.FavoritesPreviewData
import com.perrystreet.woof.presentation.favorites.uimodel.FavoriteCellUIModel
import com.perrystreet.woof.presentation.favorites.viewmodel.FavoritesViewModel
import com.perrystreet.woof.resources.R

@Composable
fun FavoritesScreen(
    state: FavoritesViewModel.State,
    onCellTap: (FavoriteCellUIModel) -> Unit,
) {
    when (state) {
        FavoritesViewModel.State.Empty -> FavoritesEmptyScreen()
        is FavoritesViewModel.State.Loaded -> FavoritesLoadedScreen(state = state, onCellTap = onCellTap)
    }
}

@Composable
private fun FavoritesEmptyScreen() {
    TemplateCenteredContent(topBar = { OrgNavigationHeaderTitle(title = stringResource(R.string.favorites_title)) }) {
        OrgEmptyState(
            title = stringResource(R.string.favorites_empty_title),
            message = stringResource(R.string.favorites_empty_message),
        )
    }
}

@Composable
private fun FavoritesLoadedScreen(
    state: FavoritesViewModel.State.Loaded,
    onCellTap: (FavoriteCellUIModel) -> Unit,
) {
    TemplateGrid(topBar = { OrgNavigationHeaderTitle(title = stringResource(R.string.favorites_title)) }) {
        items(items = state.cells, key = { cell -> cell.id }) { cell ->
            OrgPhotoCard(
                title = cell.name,
                imageState = rememberAsyncImageState(url = cell.photoUrl),
                contentDescription = stringResource(R.string.accessibility_dog_photo, cell.name),
                onTap = { onCellTap(cell) },
            )
        }
    }
}

@PreviewDevices
@Composable
private fun FavoritesScreenLoadedPreview(@PreviewParameter(ThemeProvider::class) theme: ITheme) {
    ThemedScreenPreview(theme = theme) {
        FavoritesScreen(state = FavoritesPreviewData.loaded(), onCellTap = {})
    }
}

@PreviewDevices
@Composable
private fun FavoritesScreenEmptyPreview(@PreviewParameter(ThemeProvider::class) theme: ITheme) {
    ThemedScreenPreview(theme = theme) {
        FavoritesScreen(state = FavoritesViewModel.State.Empty, onCellTap = {})
    }
}
